### PR TITLE
Proto Keywords

### PR DESCRIPTION
**Pull request by BenjaminDeleze**
This PR is linked with https://github.com/cyberbotics/webots/pull/6112

Note that the new proto removal mecanism can be improved, but it will be done in another PR.

### CODE REVIEW

File: httpdocs/ajax/proto/refresh_protos.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  require '../../../php/github_oauth.php';
+  require '../../../php/github_asset.php';
+  require '../../../php/database.php';
+  require '../../../php/create_or_update.php';
+
+  $mysqli = new mysqli($database_host, $database_username, $database_password, $database_name);
+  if ($mysqli->connect_errno)
+    error("Can't connect to MySQL database: $mysqli->connect_error");
+  $mysqli->set_charset('utf8');
+  $result = $mysqli->query('SELECT id, url FROM proto where branch="proto"') or error($mysqli->error);
```
Comment (BenjaminDeleze) [inline httpdocs/ajax/proto/refresh_protos.php:14]:
```suggestion
  $result = $mysqli->query('SELECT id, url FROM proto ') or error($mysqli->error);
```

Once reviewed and approved, we should change it such that the protos of every branches are regularly updated.
Note: as we will more than double the number of refreshed protos (from ~500 to ~1200), it is possible that we encounter some time out issue.

File: httpdocs/ajax/proto/get_keywords.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  if ($mysqli->connect_errno)
+    error("Can't connect to MySQL database: $mysqli->connect_error");
+  $mysqli->set_charset('utf8');
+  if (isset($data->id)) {
+    $query = "SELECT child.name, parent.name AS parent_name FROM proto_keywordmap LEFT JOIN proto_keyword AS child ON proto_keywordmap.keyword_id=child.keyword_id LEFT JOIN proto_keyword AS parent ON child.parent_id=parent.keyword_id WHERE proto_keywordmap.proto_id=$data->id";
+    $result = $mysqli->query($query) or error($mysqli->error);
+    $protos = array();
+    while($row = $result->fetch_array(MYSQLI_ASSOC)) {
+      array_push($protos, $row);
+    }
```
Comment (omichel) [inline httpdocs/ajax/proto/get_keywords.php]:
```suggestion
    while($row = $result->fetch_array(MYSQLI_ASSOC))
      array_push($protos, $row);
```

File: httpdocs/ajax/proto/refresh_protos.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  $continue_on_error = true;
+  while($row = $result->fetch_array(MYSQLI_ASSOC)) {
+    create_or_update_proto($row['url'], $row['id'], '');
+  }
```
Comment (omichel) [inline httpdocs/ajax/proto/refresh_protos.php]:
```suggestion
  while($row = $result->fetch_array(MYSQLI_ASSOC))
    create_or_update_proto($row['url'], $row['id'], '');
```

File: httpdocs/ajax/proto/sub_keywords.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  $mysqli->set_charset('utf8');
+  $branch = basename(dirname(__FILE__, 4));
+  if (isset($data->parent) && $data->parent!= '') {
+    $parent = $mysqli->escape_string($data->parent);
+    $query = "SELECT child.name FROM proto LEFT JOIN proto_keywordmap on proto.id = proto_keywordmap.proto_id LEFT JOIN proto_keyword AS child ON proto_keywordmap.keyword_id=child.keyword_id LEFT JOIN proto_keyword AS parent ON child.parent_id=parent.keyword_id WHERE parent.name=\"$parent\" AND branch=\"$branch\" GROUP BY proto_keywordmap.keyword_id ORDER BY COUNT(*) DESC";
+    $result = $mysqli->query($query) or error($mysqli->error);
+    $protos = array();
+    while($row = $result->fetch_array(MYSQLI_ASSOC)) {
+      array_push($protos, $row);
+    }
```
Comment (omichel) [inline httpdocs/ajax/proto/sub_keywords.php]:
```suggestion
    while($row = $result->fetch_array(MYSQLI_ASSOC))
      array_push($protos, $row);
```

File: php/create_or_update.php [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+            $no_3d_view = true;
+        } elseif (strtolower(substr($line, 0, 9)) === 'keywords:') {
+          $keywords = str_replace('keywords:', '', $line);
+          $keywords = explode(',', $keywords);
+          $kk = array();
+          foreach ($keywords as $key) {
+            $key = explode('/', $key);
+            foreach ($key as &$tag) {
+              $tag = trim($mysqli->escape_string($tag));
+            }
```
Comment (omichel) [inline php/create_or_update.php]:
```suggestion
            foreach ($key as &$tag)
              $tag = trim($mysqli->escape_string($tag));
```

File: php/create_or_update.php [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        $base_proto = $match[1];
+
+      break;
+    }
+  }
+  if (!$found_parent) {
+    $str = '';
+    for($i = 0; $i < count($externprotos); $i++) {
+      $str .= $externprotos[$i][0];
+    }
```
Comment (omichel) [inline php/create_or_update.php]:
```suggestion
    for($i = 0; $i < count($externprotos); $i++)
      $str .= $externprotos[$i][0];
```

File: php/github_asset.php [hunk 7]
```diff
@@ -49,12 +49,13 @@ function check_url($url, $proto = false) {
   return array($username, $repository, $tag_or_branch, $folder, $world_or_proto);
 }
 
+function yaml_error($msg) {
+  return "YAML file error: $msg";
+}
+
```
Comment (omichel) [inline php/github_asset.php:56]: Why this change? It increases the scope of the yaml_error function, but that doesn't seem necessary...

**Review by omichel — CHANGES_REQUESTED** [review]

File: php/github_asset.php (hunk 7 above)
Comment (BenjaminDeleze) [reply php/github_asset.php:56]: It is necessary because, when refreshing all the protos, the script calls the `github_check_yaml` function multiple times and it causes a ` Cannot redeclare yaml_error() ` error if this function is in the body of `github_check_yaml`

**Review by omichel — APPROVED** [review]
Thank you.
